azkaban-common/src/main/java/azkaban/executor/ExecutorHealthChecker.java

```diff
@@ -155,25 +165,47 @@ public void checkExecutorHealth() {
 
   /**
    * Increments executor failure count. If it reaches max failure count, sends alert emails to AZ
-   * admin.
+   * admin and executes any cleanup actions for flows on that executors.
    *
-   * @param entry executor to list of flows map entry
    * @param executor the executor
+   * @param flows flows assigned to the executor
    * @param e Exception thrown when the executor is not alive
    */
-  private void handleExecutorNotAliveCase(
-      final Entry<Optional<Executor>, List<ExecutableFlow>> entry, final Executor executor,
+  private void handleExecutorNotAliveCase(final Executor executor, final List<ExecutableFlow> flows,
       final ExecutorManagerException e) {
     logger.error("Failed to get update from executor " + executor.getId(), e);
     this.executorFailureCount.put(executor.getId(), this.executorFailureCount.getOrDefault
         (executor.getId(), 0) + 1);
-    if (this.executorFailureCount.get(executor.getId()) % this.executorMaxFailureCount == 0
-        && !this.alertEmails.isEmpty()) {
-      entry.getValue().stream().forEach(flow -> flow
-          .getExecutionOptions().setFailureEmails(this.alertEmails));
-      logger.info(String.format("Executor failure count is %d. Sending alert emails to %s.",
-          this.executorFailureCount.get(executor.getId()), this.alertEmails));
-      this.alerterHolder.get("email").alertOnFailedUpdate(executor, entry.getValue(), e);
+    if (this.executorFailureCount.get(executor.getId()) % this.executorMaxFailureCount == 0) {
+      if (!this.alertEmails.isEmpty()) {
+        logger.info(String.format("Executor failure count is %d. Sending alert emails to %s.",
+            this.executorFailureCount.get(executor.getId()), this.alertEmails));
+        this.alerterHolder.get("email")
+            .alertOnFailedExecutorHealthCheck(executor, flows, e,
+                this.alertEmails);
+      }
+      this.cleanupForMissingExecutor(executor, flows);
     }
   }
+
+  /**
+   * Perform any cleanup required for an unreachable executor.
+   *
+   * Note that ideally we would like to disable the executor such that not further executions are
+   * 'assigned' to it. However with the pull/polling based model there is currently no direct way
```

Comment:
Is there a case, that's not a not a false positive, and where an executor continues to poll even though we actually detect it to be unhealthy?

Reply:
In theory it's possible but would be quite unlikely in practice (I am not aware of an instance of this happening).
A made-up example being a temporary network issue where the web server is unable to reach an executor but both the web server as well as the executor are able to reach the mysql db. 
(The current behavior of not disabling the executor should be acceptable in those cases as well)

Reply:
Can it happen that subsequently the flow successfully finishes, and the executor updates its status?

In this case will the status go from FAILED --> SUCCESS.

Reply:
Excellent point. 
- `ExecutionControllerUtils.finalizeFlow()` includes checks for validating that the flow should not already be in a final state.
- Flow status being set to SUCCEEDED during the normal execution of the flow also checks if the the flow is already in a final state. Relevant code `FlowRunner.finalizeFlow()`

(1) Both of the above checks make it unlikely that the two path for setting in the final status will step on each other (but there is still a possible race condition as the `testAndSet` is not atomic)
(2) In addition note that executor not being reachable from healthcheck but still running is expected to be a very rare situation. 
The combined probability of (1) and (2) would make this exceedingly unlikely but the probability is still non-zero.
Do note that this is still no worse than a similar remediation being performed manually.